internal/venice-client-common/src/main/java/com/linkedin/venice/stats/VeniceOpenTelemetryDimensionsCache.java

```diff
@@ -0,0 +1,143 @@
+package com.linkedin.venice.stats;
+
+import com.linkedin.venice.exceptions.VeniceException;
+import com.linkedin.venice.stats.dimensions.VeniceMetricsDimensions;
+import com.linkedin.venice.stats.metrics.MetricEntity;
+import com.linkedin.venice.utils.concurrent.VeniceConcurrentHashMap;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * Cache for storing dimensions for OpenTelemetry metrics. The cache is used to avoid churning
+ * same dimension objects every time a metric is emitted potentially affecting GC.
+ *
+ * This class holds
+ * 1. some base dimensions to be reused
+ * 2. a Thread-local map to pass in the dimension and its values to create {@link Attributes}
+ *    and avoid creating temp maps/arrays
+ * 3. the actual dimensions cache to hold the {@link Attributes} for every cache key
+ *
+ * Right now, this class is structured to be instantiated more than once for the same metric repository
+ * based on having the base dimensions as part of the stats objects created for different request types,
+ * to play along with the current design and pre create the base dimensions and its key. Eventually this
+ * might come to a point where there are no base dimensions based on the restructure when removing
+ * tehuti and this can be made a static class as part of {@link VeniceOpenTelemetryMetricsRepository}
+ */
+public class VeniceOpenTelemetryDimensionsCache {
+  private static final Logger LOGGER = LogManager.getLogger(VeniceOpenTelemetryDimensionsCache.class);
+  private final VeniceOpenTelemetryMetricsRepository otelRepository;
+  /**
+   * Store the base dimensions: created once per request-type, per store, etc. metrics
+   * extending {@link AbstractVeniceHttpStats} where the common dimensions can be reused
+   */
+  private final Attributes baseMetricDimensions;
+  /** part of cache key from the base dimensions: To reuse */
+  private final Set<VeniceMetricsDimensions> baseMetricDimensionsSet;
+  /** the keys in the base dimensions to ignore when creating the cache key from all other dimensions */
+  private final String baseMetricDimensionsKey;
+
+  /** used to pass in the dimension and its values to create {@link Attributes} and avoid creating temp maps/arrays */
+  private static final ThreadLocal<Map<VeniceMetricsDimensions, String>> threadLocalReusableDimensionsMap =
```

Comment:
Is this correct that, ThreadLocal requires every router thread servicing a client query have to create a copy of this map? If it is the case, do we know the upper limit of how large this number could be?